### ISSUE TITLE
data integrity checks for uploads/downloads

### ISSUE THREAD

**Issue by silvolu** [GoogleCloudPlatform/google-cloud-python#547]
For uploads/downloads, check in order of priority:
- md5 computed for local file against md5 contained in Object’s metadata (more expensive, more reliable)
- crc32c computed for local file against crc32c contained in Object’s metadata (less expensive, less reliable)

[eaxmple gslib code for the checks chain](https://github.com/GoogleCloudPlatform/gsutil/blob/ccca18e64f3f7d87c989675f8673af7bdecae0d1/gslib/copy_helper.py#L805)

For copies of objects already in the storage, compare the checksums in the Object’s metadata with the same priority (md5 could not be available in some cases)
[example gslib_code](https://github.com/GoogleCloudPlatform/gsutil/blob/ccca18e64f3f7d87c989675f8673af7bdecae0d1/gslib/copy_helper.py#L756)

If the checksum check fails, the created file should be deleted by the client, and an exception raised.
[example gslib_code](https://github.com/GoogleCloudPlatform/gsutil/blob/ccca18e64f3f7d87c989675f8673af7bdecae0d1/gslib/copy_helper.py#L1694)

We should allow to override the configuration for the checks (e.g. use only crc32c, do not perform checks), but it should be strongly discouraged.

**Comment by lukesneeringer** [comment]
This seems to have been done.

**Comment by dhermes** [comment]
@lukesneeringer Nope, that was the CRC-32 PR that you closed. We also need the option of an MD5 integrity check.

**Comment by chemelnucfin** [comment]
I'm going to go ahead and close this and put this into our feature backlog.  Please feel free to reopen if there are any concerns.
